### PR TITLE
!typ #22970 Actor.restarter renamed to supervise

### PR DESCRIPTION
**Pull request by ktoso**
SOME TESTS ARE FAILING STILL - Will finish working on this after lunch.

Updated supervision API for Typed, with the following benefits:

- reads nicer: was `Actor.restarter[Something](...)` now is: `Actor.supervise(behavior).onFailure(howToDealWithIt)`

No default is picked, this has the benefit of:

- exposing people to the fact there's more to life than "restart"
  - esp interesting since now we can just use backoff supervision in those
- Java and Scala DSL look the same

### CODE REVIEW

File: akka-typed/project/build.properties [hunk 1]
```diff
@@ -0,0 +1 @@
+sbt.version=0.13.15
```
Comment (johanandren) [inline akka-typed/project/build.properties]: wut?!

Comment (ktoso) [inline akka-typed/project/build.properties]: Oh, weird - will remove ofc.

**Review by patriknw — COMMENTED** [review]
I haven't looked at all details yet, but what I can see looks promising.

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 2]
```diff
@@ -218,17 +216,25 @@ object Actor {
...
-  final class Restarter[Thr <: Throwable: ClassTag](c: ClassTag[Thr], strategy: SupervisorStrategy) {
-    def wrap[T](b: Behavior[T]): Behavior[T] = akka.typed.internal.Restarter(Behavior.validateAsInitial(b), strategy)(c)
+  def supervise[T](wrapped: Behavior[T]): Supervise[T] =
+    new Supervise(wrapped)
+
+  final class Supervise[T](val wrapped: Behavior[T]) extends AnyVal {
+    /** Specify the [[SupervisorStrategy]] to be invoked when the wrapped behaior throws. */
+    def onFailure[Thr <: Throwable: ClassTag](strategy: SupervisorStrategy): Behavior[T] = {
+      val tag = implicitly[ClassTag[Thr]]
+      val effectiveTag = if (tag == ClassTag(classOf[Nothing])) ClassTag(classOf[Throwable]) else tag
```
Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: This is sneaky :-/
I'm open to any other ideas how to solve this. The problem is: sadly when someone would write `.onFailure(blabla)` (missing `[...]`), then Nothing would get inferred so we're attempting to match if a thing is Nothing, and not Throwable in the internals then.

I tried some type level hackery sadly that's not working : https://stackoverflow.com/questions/27912014/enforce-a-lower-bounds-that-excludes-a-super-type

Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: I think that is a good way. You should probably cache the ClassTag-Nothing

Comment (ktoso) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: Cool, if we think that's fine then I'll just cache it and it'll work fine :)

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 3]
```diff
@@ -218,17 +216,26 @@ object Actor {
...
-    def wrap[T](b: Behavior[T]): Behavior[T] = akka.typed.internal.Restarter(Behavior.validateAsInitial(b), strategy)(c)
+  def supervise[T](wrapped: Behavior[T]): Supervise[T] =
+    new Supervise(wrapped)
+
+  private final val NothingClassTag = ClassTag(classOf[Nothing])
+  final class Supervise[T](val wrapped: Behavior[T]) {
+    /** Specify the [[SupervisorStrategy]] to be invoked when the wrapped behaior throws. */
+    def onFailure[Thr <: Throwable: ClassTag](strategy: SupervisorStrategy): Behavior[T] = {
+      val tag = implicitly[ClassTag[Thr]]
+      val effectiveTag = if (tag == NothingClassTag) ClassTag(classOf[Throwable]) else tag
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: `Throwable` tag should also be cached

File: akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala [hunk 4]
```diff
@@ -218,17 +216,26 @@ object Actor {
...
    * resume, backoff.
    *
+   * Note that only [[scala.util.control.NonFatal]] throwables will trigger the supervision strategy.
+   *
    * Example:
    * {{{
    * val dbConnector: Behavior[DbCommand] = ...
-   * val dbRestarts = Restarter[DbException]().wrap(dbConnector)
+   * val dbRestarts = supervise(dbConnector).onFailure(SupervisorStrategy.resume) // handle all NonFatal exceptions
+   * val dbSpecificRestarts = supervise(dbConnector).onFailure[IndexOutOfBoundsException](SupervisorStrategy.resume) // handle all NonFatal exceptions
```
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/scaladsl/Actor.scala]: use `Actor.supervise` in examples

**Review by patriknw — CHANGES_REQUESTED** [review]
looking great, one minor thing

also nice to see that we were able to change without touching the implementation, just as a confirmation that we did something right with the separation of api and impl

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala [hunk 5]
```diff
@@ -227,12 +227,29 @@ object Actor {
...
    */
-  def restarter[T, Thr <: Throwable](
-    clazz:           Class[Thr],
-    strategy:        SupervisorStrategy,
-    initialBehavior: Behavior[T]): Behavior[T] = {
-    Restarter(Behavior.validateAsInitial(initialBehavior), strategy)(ClassTag(clazz))
+  def supervise[T](wrapped: Behavior[T]): Supervise[T] =
+    new Supervise[T](wrapped)
+
+  final class Supervise[T](wrapped: Behavior[T]) {
```
Comment (johanandren) [inline akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: Make the constructor private/internal to give us some more bin comp freedom?

**Review by johanandren — APPROVED** [review]
LGTM

**Comment by johanandren** [comment]
I think we should update that supervision blog article with a note about the renaming as well when this is released.

**Comment by patriknw** [comment]
yes, I'll update the article and the example repo

File: akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala (hunk 5 above)
Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: +1 but it will not be visible from java

Comment (patriknw) [inline akka-typed/src/main/scala/akka/typed/javadsl/Actor.scala]: I mean, it will still show up as public from java (as usual)

**Review by johanandren — APPROVED** [review]
LGTM!

**Review by patriknw — APPROVED** [review]
LGTM

**Comment by patriknw** [comment]
Refs #22970

**Pull request merged**